packages/tpetra/core/src/Tpetra_Details_unpackCrsMatrix.hpp

```diff
@@ -70,270 +72,345 @@ class Distributor;
+    bad_index(Tpetra::Details::OrdinalTraits<LO>::invalid()),
+    num_ent(0),
+    offset(0),
+    expected_num_bytes(0),
+    num_bytes(0),
+    buf_size(0),
+    wrong_num_bytes_error(false),
+    out_of_bounds_error(false) {}
+
+  bool success() const
+  {
+    // Any possible error would result in bad_index being changed from
+    // `invalid` to the index associated with the error.
+    return bad_index == Tpetra::Details::OrdinalTraits<LO>::invalid();
   }
-  else if (combineMode == REPLACE) {
-    for (int k=0; k<numEnt; k++) {
-      LO lclColInd = lclColMap.getLocalElement(cols[k]);
-      numValid += lclMatrix.replaceValues(lclRow, &lclColInd, 1, &vals[k], false, atomic);
+
+  std::string summary() const
+  {
+    std::ostringstream os;
+    if (wrong_num_bytes_error || out_of_bounds_error) {
+      if (wrong_num_bytes_error) {
+        os << "At index i = " << bad_index
+           << ", expectedNumBytes > numBytes.";
+      }
+      else if (out_of_bounds_error) {
+        os << "First invalid offset into 'imports' "
+           << "unpack buffer at index i = " << bad_index << ".";
+      }
+      os << "  importLIDs[i]: " << bad_index
+         << ", bufSize: " << buf_size
+         << ", offset: " << offset
+         << ", numBytes: " << num_bytes
+         << ", expectedNumBytes: " << expected_num_bytes
+         << ", numEnt: " << num_ent;
     }
+    return os.str();
   }
-  else if (combineMode == ABSMAX) {
-    TEUCHOS_TEST_FOR_EXCEPTION(
-        true, std::invalid_argument,
-        "ABSMAX combine mode is not yet implemented for a matrix that has a "
-        "static graph (i.e., was constructed with the CrsMatrix constructor "
-        "that takes a const CrsGraph pointer).");
-  }
-  else if (combineMode == INSERT) {
-    TEUCHOS_TEST_FOR_EXCEPTION(
-        true, std::invalid_argument,
-        "INSERT combine mode is not allowed if the matrix has a static graph "
-        "(i.e., was constructed with the CrsMatrix constructor that takes a "
-        "const CrsGraph pointer).");
+};
+
+
+/// \brief Unpacks and combines a single row of the CrsMatrix.
+///
+/// Data (bytes) describing the row of the CRS matrix are "unpacked"
+/// from a single (concatenated) char* in to the row of the matrix
+///
+/// \tparam LocalMatrixType the specialization of the KokkosSparse::CrsMatrix
+///   local matrix
+/// \tparam LocalMapType the type of the local column map
+template<class CountType, class OffsetType, class LocalMatrixType, class LocalMapType>
+struct UnpackCrsMatrixAndCombineFunctor {
+
+  typedef CountType count_type;
+  typedef OffsetType offset_type;
+  typedef LocalMatrixType local_matrix_type;
+  typedef LocalMapType local_map_type;
+  typedef typename local_matrix_type::value_type IST;
+  typedef typename local_matrix_type::ordinal_type LO;
+  typedef typename local_map_type::global_ordinal_type GO;
+
+  typedef Kokkos::View<const count_type*, Kokkos::HostSpace, Kokkos::MemoryUnmanaged> num_packets_per_lid_view_type;
```

Comment:
I can guess why you made these all HostSpace Views; it's because the original data come in as host Teuchos::ArrayView.  However, none of these host Views are accessible from the GPU.  We will need to copy those data to the GPU before we may access them there.  A better approach would be to change CrsMatrix to implement the "new DistObject interface."  That way, those data would come in as device Views already, so we wouldn't have to worry about copying them.